examples/base_api.py

```diff
@@ -0,0 +1,86 @@
+"""A run through of base API operations on all supported clouds."""
+
+import os
+from contextlib import suppress
+
+import pycloudlib
+from pycloudlib.cloud import BaseCloud
+
+
+cloud_config = """#cloud-config
+runcmd:
+  - echo 'hello' > /home/ubuntu/example.txt
+"""
+
+
+def exercise_api(client: BaseCloud):
+    """Run through supported functions in the base API."""
+    try:
+        image_id = client.released_image('focal')
+    except NotImplementedError:
+        image_id = client.daily_image('focal')
+    print('focal image id: {}'.format(image_id))
+    print('launching instance...')
+    instance = client.launch(
+        image_id=image_id,
+        user_data=cloud_config
+    )
+    assert repr(instance) == repr(client.get_instance(instance.name))
+
+    print('instance name: {}'.format(instance.name))
+    with suppress(NotImplementedError):
+        print('instance ip: {}'.format(instance.ip))
+
+    print('starting instance...')
+    instance.start()
+    print('waiting for cloud-init...')
+    instance.execute('cloud-init status --wait --long')
+    with suppress(NotImplementedError):
+        instance.console_log()
+    assert instance.execute('cat /home/ubuntu/example.txt').stdout == 'hello'
+
+    print('restarting instance...')
+    instance.execute('sync')  # Prevent's some wtfs :)
+    instance.restart()
+    assert instance.execute('cat /home/ubuntu/example.txt').stdout == 'hello'
+
+    print('shutting down instance...')
+    instance.shutdown()
+    print('starting instance...')
+    instance.start()
+    print(instance.execute('cat /home/ubuntu/example.txt').stdout)
+    snapshot_id = None
+    with suppress(NotImplementedError):
+        print('snapshotting instance...')
+        snapshot_id = client.snapshot(instance)
+        print('snapshot image id: {}'.format(snapshot_id))
+    if snapshot_id:
+        assert snapshot_id != image_id
+        print('deleting snapshot...')
+        client.delete_image(snapshot_id)
+
+    print('deleting instance...')
+    instance.delete()
+
```

Comment:
Maybe we could further improve the test by launching a vm from the `snapshot_id`